SubsoniciousKit/SubsoniciousKit/Network/Request/NetworkRequest.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  NetworkRequest.swift
+//  SubsoniciousKit
+//
+//  Created by Bilal on 05/07/2020.
+//  Copyright © 2020 Bilal Benlarbi. All rights reserved.
+//
+
+import CryptoKit
+import Foundation
+
+class NetworkRequest {
```

Comment:
with a baseURL URL typed, you don't have to worry about some of your parameters :

```
class NetworkRequest {
    let server: Server

    init(server: Server) {
        self.server = server
    }

    var path: String { "" }

    var queryItems: [URLQueryItem] {
        [URLQueryItem(name: ParameterName.version, value: ParameterValue.version),
         URLQueryItem(name: ParameterName.clientApplication, value: ParameterValue.clientApplication),
         URLQueryItem(name: ParameterName.dataFormat, value: ParameterValue.dataFormat),
         URLQueryItem(name: ParameterName.username, value: username),
         URLQueryItem(name: ParameterName.authenticationToken, value: authenticationToken),
         URLQueryItem(name: ParameterName.salt, value: salt)]
    }

    var url: URL? {
        guard var urlComponents = URLComponents(url: server.baseURL, resolvingAgainstBaseURL: true) else { return nil }

        urlComponents.path = path
        urlComponents.queryItems = queryItems

        return urlComponents.url
    }

    private lazy var username: String = {
        server.username
    }()

    private lazy var authenticationToken: String = {
        let saltedPassword = server.password + salt
        let token = try? md5(saltedPassword)
        return token ?? ""
    }()

    private lazy var salt: String = {
        randomString(length: 8)
    }()
}
```

Reply:
Or you could create a URLComponents directly from the string `URLComponents(string: server.baseURL)`

Reply:
here is a more compact of your request : 
```
struct NetworkRequest {
    private var urlComponents: URLComponents

    init(server: Server, path: String) throws {
        guard let components = URLComponents(url: server.baseURL, resolvingAgainstBaseURL: true) else {
            throw Error.badURL(server.baseURL)
        }
        self.urlComponents = components

        let salt: String = randomString(length: 8)
        let authenticationToken = try md5(server.password + salt)

        urlComponents.path = path
        urlComponents.queryItems = [
            URLQueryItem(name: ParameterName.version, value: ParameterValue.version),
            URLQueryItem(name: ParameterName.clientApplication, value: ParameterValue.clientApplication),
            URLQueryItem(name: ParameterName.dataFormat, value: ParameterValue.dataFormat),
            URLQueryItem(name: ParameterName.username, value: server.username),
            URLQueryItem(name: ParameterName.authenticationToken, value: authenticationToken),
            URLQueryItem(name: ParameterName.salt, value: salt)
        ]
    }

    var url: URL? { urlComponents.url }
}
```

Reply:
Good idea, but that's not conform with inheritance unfortunately :(
Also I prefer keeping the baseURL as a String to use the errors from the URLComponents when trying to build the url

Reply:
As you wish. I strongly disagree with you for this, (no inheritance could be better for me) and there is no error handling from the URLComponents itself, so in all cases we need to handle it. It's your app so you do what you want, but I dislike this ^^

Reply:
I answered here after : https://github.com/bilalb/Subsonicious/pull/5/files#r456795164

This conversation didn't show during the review..

Reply:
Resolving the discussion as per my previous comment from another discussion about the same topic :)